Classes/Promise.swift

```diff
@@ -89,32 +71,23 @@ extension Array {
 public class PromiseCancelToken {
     
     public typealias Canceled = () -> Void
-    
-    internal let cancelFlag: AtomicCancel = AtomicCancel()
+    var workItem: DispatchWorkItem? = DispatchWorkItem(block: {}) // we can't inherit so we will use Decorator Pattern instead
     static let queue = DispatchQueue(label: "com.github.benski.promise")
-    var onCancel : Canceled?
-    
+    internal var isCanceled : Bool { return workItem?.isCancelled ?? false}
+    internal var cancelBlocks = [DispatchWorkItem]()
+
     public func cancelled(_ onCancel: @escaping Canceled, on queue: DispatchQueue) {
-        let wrappedBlock = {
-            queue.async {
-                onCancel()
-            }
-        }
-        
-        PromiseCancelToken.queue.async {
-            if self.cancelFlag.isCanceled {
-                wrappedBlock()
-            } else {
-                self.onCancel = wrappedBlock
-            }
+        if let workItem = workItem {
+            let cancelItem = DispatchWorkItem(block: onCancel)
```

Comment:
the same here

Reply:
inheritQoS would be better don't you think

Reply:
as far as I understand these very similar flags they increase priority for tasks:
```Swift
static let enforceQoS: DispatchWorkItemFlags
//Prefer the quality-of-service class associated with the block.
static let inheritQoS: DispatchWorkItemFlags
//Prefer the quality-of-service class associated with the current execution context.
```
Here we work with blocks then we should use enforceQoS. It's my point.

Reply:
nah, it shouldn't raise the priority of the queue and indirectly other tasks? if its inherit, then its own qos can be raised to a higher one if the queue specifies it. But otherwise no.